scheduler/src/cook/tools.clj

```diff
@@ -951,6 +952,39 @@
                 [usage' (below-quota? quota usage')]))]
       (filter-sequential filter-with-quota usage queue))))
 
+
+; This is used by the /unscheduled_jobs code to determine whether
+; or not to report rate-limiting as a reason for being pending
+(defonce pool->user->num-rate-limited-jobs (atom {}))
+
+(defn filter-pending-jobs-for-ratelimit
+  [pool-name queue]
+  (let [enforcing-job-launch-rate-limit? (ratelimit/enforce? quota/per-user-per-pool-launch-rate-limiter)
+        user->number-jobs (atom {})
+        user->rate-limit-count (atom {})
+        user-within-launch-rate-limit?-fn
+        (fn
+          [{:keys [job/user]}]
+          ; Account for each time we see a job for a user.
+          (let [token-key (quota/->token-key pool-name user)
+                _ (swap! user->number-jobs update token-key #(inc (or % 0)))
+                tokens-left (ratelimit/get-token-count! quota/per-user-per-pool-launch-rate-limiter token-key)
+                number-jobs-for-user-so-far (@user->number-jobs token-key)
+                is-rate-limited? (> number-jobs-for-user-so-far tokens-left)]
+            (when is-rate-limited?
+              (swap! user->rate-limit-count update token-key #(inc (or % 0))))
+            (not (and is-rate-limited? enforcing-job-launch-rate-limit?))))
+        filtered-queue (doall (filter user-within-launch-rate-limit?-fn queue))]
```

Comment:
We can't `doall` here because we'd be realizing the entire queue of pending jobs, which could be hundreds of thousands of jobs in some environments. We need to let the laziness keep flowing.

Reply:
Looks like you've now implemented that, correct?

Reply:
Yup.